### DISCUSSION TITLE
[Question][Lecture6] Dealing with variant of interpretation

### DISCUSSION THREAD

**Discussion by Calmadeas** [prosyslab-classroom/cs424-program-reasoning#369] [Q&A]
In class, while proving the validity of
$F: a[i] = e \rightarrow \forall j. a \langle i \triangleleft e \rangle [j] = a[j]$,
we reached
$(1) \ I \vDash a[i] = e$, $(2) \ I' \nvDash a \langle i \triangleleft e \rangle [j] = a[j]$ and $(3) \ I' \vDash i = j$ where $I' = I \triangleleft \\{ j \mapsto v \\}$.
This is enough if we can say
$(1)$ implies $(4) \ I' \vDash a[i] = e$,
since $(3)$ and $(4)$ implies $(5) \ I' \vDash a \langle i \triangleleft e \rangle [j] = a[j]$ which is contradictory to $(2)$.

The only problem is that I couldn't find a rule that implies $(4)$ from $(1)$.
Here are the alternative answers I found:
a) Add a new rule.
$I \vDash F$ implies $I' \vDash F$ where $I'$ is a $x$-variant of $I$ and $x$ is free in $F$.

b) Move $\forall j.$ to the outmost. That is, replace $F$ by
$F':  \forall j.(a[i] = e \rightarrow a \langle i \triangleleft e \rangle [j] = a[j])$.
Then we are happy since we will get $(4)$ instead of $(1)$ from the beginning of the proof.

c) There is a rule but I just couldn't find.

So what would be the correct way?

**Comment by KihongHeo** [comment]
Hi. Good point.

Your option `a` is (almost) correct. But we do not need an additional rule. It is trivial according to the semantics of FOL. See page 13 in lecture5.pdf.
$I \models F$ means that $F$ is already true under $I$. Any additional interpretation of unused symbols ( $j$ in your example) does not change the result.
